### ISSUE TITLE
Quark Catalyst V3 & BeagleWrite crash

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#400]
Reported by Marco Verpelli, on cea2 "Quark Catalyst V3 crash with AppleWin 1.26.2.0 & 1.26.2.1":

I am using the LoGo version available at
http://www.brutaldeluxe.fr/crack/quark_catalyst_v3_logo.po

Add to the list: Beaglewrite.
Perhaps something to do with interrupt and mouse 

> NB. These versions are both experimental. 

I know, and continue to use 1.26.1.1.
If you turn off mouse interface in slot 4, both program starts.
No problems with: Mouse Paint, Geos, DazzleDraw, Mouse Desktop Toolkit, Mouse Calc, Instant Pascal and Apple II Desktop.

**Comment by tomcw** [comment]
I can repro with BeagleWriteP8.po (800KB file).

I have isolated the change to the `MemSetPaging()` change at 87862bcf28d32ed8f9a4e9d0433550781193291b.

**Comment by tomcw** [comment]
In `MemSetPaging()`, if I ignore the `write` flag (so do the same for a write as I'm already doing for a read) then Quark Catalyst works (but a2audit v1.04 fails the LC test).

**Comment by tomcw** [comment]
This seems to fix it (but only tested Quark Catalyst, BeagleWrite & a2audit):

Changed from:
``` C++
		if (!write)	// GH#392
		{
			SetMemMode(memmode & ~(MF_WRITERAM));
			BOOL bWriteRam = (address & 1);
			if (g_bLastWriteRam && bWriteRam)
				SetMemMode(memmode | MF_WRITERAM);
			g_bLastWriteRam = bWriteRam;
		}
		else
		{
			g_bLastWriteRam = 0;
		}
```

to:
``` C++
		SetMemMode(memmode & ~(MF_WRITERAM));
		BOOL bWriteRam = (address & 1);
		if (g_bLastWriteRam && bWriteRam)
			SetMemMode(memmode | MF_WRITERAM);

		if (!write)	// GH#392
			g_bLastWriteRam = bWriteRam;
		else
			g_bLastWriteRam = 0;
```

**Comment by tomcw** [comment]
Hmm... but the above patch doesn't allow Gemstone Warrior to boot (#395).

**Comment by sicklittlemonkey** [comment]
Yes, was the bug I saw - some of that behaviour occurs on a read. Will have a look tonight.

**Comment by sicklittlemonkey** [comment]
I don't have time to check it in, but the following should be right.
Note: there are some errors in Sather's book, but he restates things enough to find them.

```
		if (address & 1)	// GH#392
		{
			if (!write && g_bLastWriteRam)
			{
				SetMemMode(memmode | MF_WRITERAM); // UTAIIe:5-23
			}
		}
		else
		{
			SetMemMode(memmode & ~(MF_WRITERAM)); // UTAIIe:5-23
		}
		g_bLastWriteRam = (address & 1) && (!write); // UTAIIe:5-23

```

**Comment by tomcw** [comment]
Thanks for digging into this.
I won't have time to check this until next week.

**Comment by sicklittlemonkey** [comment]
Tested Quark Catalyst, BeagleWrite and Gemstone Warrior.

**Comment by peterferrie** [comment]
It's not clear to me that this is completely correct, but I didn't find anything wrong with Sather.
You might check src\devices\bus\a2bus\ramcard16k.cpp in MAME to see if the implementation matches.

**Comment by sicklittlemonkey** [comment]
They look equivalent to me, but the LC logic sure is fun to lose your head
in for an hour.

I even created a handy LibreOffice spreadsheet so I could sort by each
signal to check the logic groupings as I looked at different versions of
the code.

HRAMRD can be calculated in a line or two, or you can find a reason to do
it in a 16 line switch as in MAME. ; - )

Errors in Sather on this:
Table 5-4 on page 5-22 has HRAMRD logic equations using A1 and A2 instead
of A0 and A1.
Figure 5.13a on page 5-30 has BANK1 fed A2 instead of A3.

The numbered description on page 5-23 is good, as is Table 5.5.

Cheers,
Nick.

**Comment by sicklittlemonkey** [comment]
Here's the language card logic spreadsheet as ODS and XSLX created by LibreOffice.

(I did attach it to my previous email, but GitHub discarded it. In fact it doesn't support attaching ODS, and doesn't recognize the LibreOffice created XSLX to allow it. So ZIP it is.)

[LanguageCardSpreadsheet.zip](https://github.com/AppleWin/AppleWin/files/907441/LanguageCardSpreadsheet.zip)

**Comment by Michaelangel007** [comment]
Nick, I appreciate you including both the LibreOffice and Excel file! Thanks!

Here it is in plain text -- well markup ;-) for future reference.

|Address|Bits|R/W|Bank|HRAMRD|PREWRITE|HRAMWRT|
|:------|---:|:-:|:--:|:----:|:------:|:-----:|
|C080|0000|	R|	2|	1|	0|	0 |
|C080|0000|	W|	2|	1|	0|	0 |
|C081|0001|	R|	2|	0|	1|	1*|
|C081|0001|	W|	2|	0|	0|	- |
|C082|0010|	R|	2|	0|	0|	0 |
|C082|0010|	W|	2|	0|	0|	0 |
|C083|0011|	R|	2|	1|	1|	1*|
|C083|0011|	W|	2|	1|	0|	- |
|C084|0100|	R|	2|	1|	0|	0 |
|C084|0100|	W|	2|	1|	0|	0 |
|C085|0101|	R|	2|	0|	1|	1*|
|C085|0101|	W|	2|	0|	0|	- |
|C086|0110|	R|	2|	0|	0|	0 |
|C086|0110|	W|	2|	0|	0|	0 |
|C087|0111|	R|	2|	1|	1|	1*|
|C087|0111|	W|	2|	1|	0|	- |
|C088|1000|	R|	1|	1|	0|	0 |
|C088|1000|	W|	1|	1|	0|	0 |
|C089|1001|	R|	1|	0|	1|	1*|
|C089|1001|	W|	1|	0|	0|	- |
|C08A|1010|	R|	1|	0|	0|	0 |
|C08A|1010|	W|	1|	0|	0|	0 |
|C08B|1011|	R|	1|	1|	1|	1*|
|C08B|1011|	W|	1|	1|	0|	- |
|C08C|1100|	R|	1|	1|	0|	0 |
|C08C|1100|	W|	1|	1|	0|	0 |
|C08D|1101|	R|	1|	0|	1|	1*|
|C08D|1101|	W|	1|	0|	0|	- |
|C08E|1110|	R|	1|	0|	0|	0 |
|C08E|1110|	W|	1|	0|	0|	0 |
|C08F|1111|	R|	1|	1|	1|	1*|
|C08F|1111|	W|	1|	1|	0|	- |

**Comment by tomcw** [comment]
Thanks Nick for reworking & fixing the LC code.
I've built, tested and pre-released a new AppleWin 1.26.2.2 version [here](https://github.com/AppleWin/AppleWin/releases/tag/v1.26.2.2).
